piptools/scripts/compile.py

```diff
@@ -421,7 +427,13 @@ def cli(
 
     # Filter out pip environment markers which do not match (PEP496)
     constraints = [
-        req for req in constraints if req.markers is None or req.markers.evaluate()
+        req
+        for req in constraints
+        if req.markers is None
+        # We explicitly set extra=None to filter out optional requirements
+        # since evaluating an extra marker with no environment raises UndefinedEnvironmentName
+        # (see https://packaging.pypa.io/en/latest/markers.html#usage)
+        or req.markers.evaluate(dict(extra=None))
```

Comment:
Probably, it's good to make it possible to specify extra via a CLI argument. However, it's not a big deal for the current PR since, as far as I see, it wasn't supported by the previous implementation.

Reply:
Yeah, thanks for pointing that out. BTW, the discussion about setuptools extras is here: https://github.com/jazzband/pip-tools/issues/908#issuecomment-540687007.